source/slang/glsl.meta.slang

```diff
@@ -6803,3 +6803,720 @@ __glsl_version(430) [require(glsl)]
+
+[require(hlsl, sm_5_0)]
+__glsl_version(430) [require(glsl)]
+__spirv_version(1.0) [require(spirv)]
+[ForceInline]
+public void memoryBarrier()
+{
+    __target_switch
+    {
+    case hlsl: __intrinsic_asm "AllMemoryBarrier()";
+    case glsl: __intrinsic_asm "memoryBarrier()";
+    case spirv:
+    {
+        spirv_asm
+        {
+            OpCapability Shader;
+            OpMemoryBarrier Invocation AcquireRelease|WorkgroupMemory|ImageMemory|UniformMemory
+        };
+    }
+    }
+}
+// Does not exist in vulkan. The equivlent is now `memoryBarrierBuffer`
+// for vulkan since atomic_counter is a wrapped storage buffer object
+[require(hlsl, sm_5_0)]
+__glsl_version(430) [require(glsl)]
+__spirv_version(1.0) [require(spirv)]
+[ForceInline]
+public void memoryBarrierAtomicCounter()
+{
+    __target_switch
+    {
+    case hlsl: __intrinsic_asm "AllMemoryBarrier()";
+    case glsl: __intrinsic_asm "memoryBarrierBuffer()";
+    case spirv:
+    {
+        spirv_asm
+        {
+            OpCapability Shader;
+            OpMemoryBarrier Invocation AcquireRelease|UniformMemory
+        };
+    }
+    }
+}
+[require(hlsl, sm_5_0)]
+__glsl_version(430) [require(glsl)]
+__spirv_version(1.0) [require(spirv)]
+[ForceInline]
+public void memoryBarrierBuffer()
+{
+    __target_switch
+    {
+    case hlsl: __intrinsic_asm "DeviceMemoryBarrier()";
+    case glsl: __intrinsic_asm "memoryBarrierBuffer()";
+    case spirv:
+    {
+        spirv_asm
+        {
+            OpCapability Shader;
+            OpMemoryBarrier Invocation AcquireRelease|UniformMemory
+        };
+    }
+    }
+}
+[require(hlsl, sm_5_0)]
+__glsl_version(430) [require(glsl)]
+__spirv_version(1.0) [require(spirv)]
+[ForceInline]
+public void memoryBarrierShared()
+{
+    __target_switch
+    {
+    case hlsl: __intrinsic_asm "GroupMemoryBarrier()";
+    case glsl: __intrinsic_asm "memoryBarrierShared()";
+    case spirv:
+    {
+        spirv_asm
+        {
+            OpCapability Shader;
+            OpMemoryBarrier Workgroup AcquireRelease|WorkgroupMemory
+        };
+    }
+    }
+}
+[require(hlsl, sm_5_0)]
+__glsl_version(430) [require(glsl)]
+__spirv_version(1.0) [require(spirv)]
+[ForceInline]
+public void memoryBarrierImage()
+{
+    __target_switch
+    {
+    case hlsl: __intrinsic_asm "DeviceMemoryBarrier()";
+    case glsl: __intrinsic_asm "memoryBarrierImage()";
+    case spirv:
+    {
+        spirv_asm
+        {
+            OpMemoryBarrier Invocation AcquireRelease|ImageMemory
+        };
+    }
+    }
+}
+[require(hlsl, sm_5_0)]
+__glsl_version(430) [require(glsl)]
+__spirv_version(1.0) [require(spirv)]
+[ForceInline]
+public void groupMemoryBarrier()
+{
+    __target_switch
+    {
+    case hlsl: __intrinsic_asm "GroupMemoryBarrier()";
+    case glsl: __intrinsic_asm "groupMemoryBarrier()";
+    case spirv:
+    {
+        spirv_asm
+        {
+            OpMemoryBarrier Workgroup AcquireRelease|WorkgroupMemory|ImageMemory|UniformMemory
```

Comment:
Workgroup storage class objects are shared memory objects. 

glslang uses this flag for shared memory syncs (I do not have the link on me currently) & [dxc uses it for shared memory sync's](https://github.com/microsoft/DirectXShaderCompiler/blob/122464aeefd04553b635e167b193a10cd316f140/tools/clang/test/CodeGenSPIRV/intrinsics.groupmemorybarrier.hlsl)

Reply:
`Control the ordering of all memory transactions
issued within a single shader invocation, as
viewed by other invocations in the same
workgroup.
Only available in compute shaders`

Reply:
let me see what shader-playground compiles the function as

Reply:
compiles `               OpMemoryBarrier %uint_2 %uint_3400`
2 = WorkGroup Memory scope
3400 = all memory

Reply:
0x8|0x40|0x100|0x400|0x800

Acquire Release | uniform | workgroup | atomic_counter|imageMemory

hmm, looks like AtomicCounter Memory is used

Reply:
`
Validation: error 0: Validation Error: [ VUID-VkShaderModuleCreateInfo-pCode-08737 ] | MessageID = 0xa5625282 | vkCreateShaderModule(): pCreateInfo->pCode (spirv-val produced an error):
Capability AtomicStorage is not allowed by Vulkan 1.2 specification (or requires extension)
  OpCapability AtomicStorage
. The Vulkan spec states: If pCode is a pointer to SPIR-V code, pCode must adhere to the validation rules described by the Validation Rules within a Module section of the SPIR-V Environment appendix (https://vulkan.lunarg.com/doc/view/1.3.275.0/windows/1.3-extensions/vkspec.html#VUID-VkShaderModuleCreateInfo-pCode-08737)

Validation: error 0: Validation Error: [ VUID-VkShaderModuleCreateInfo-pCode-08739 ] | MessageID = 0x605314fa | vkCreateComputePipelines(): pCreateInfos[0].stage SPIR-V has Capability (Unhandled OpCapability) declared, but this is not supported by Vulkan. The Vulkan spec states: If pCode is a pointer to SPIR-V code, pCode must not declare any capability that is not supported by the API, as described by the Capabilities section of the SPIR-V Environment appendix (https://vulkan.lunarg.com/doc/view/1.3.275.0/windows/1.3-extensions/vkspec.html#VUID-VkShaderModuleCreateInfo-pCode-08739)
`
well, whatever was outputted by glslang was clearly incorrect for GLSL, does not work in (modern) vulkan

Reply:
I don't know why you are talking about AtomicCounterMemory here.
But I take back what I said.
I think "groupMemoryBarrier()" should keep WorkgrouMemory.

Reply:
I brought up `AtomicCounterMemory` since I decided to look at what glslang produces for glsl->SPIR-V to see what is the correct answer

and it turns out glslang produces incorrect SPIR-V which really should not work